File: plugin/octo_beat_plugin/ios/Classes/core/device/handler/tasks/notify/NotifyECGDataHandler.swift
//
//  NotifyECGDataHandler.swift
//  OctoBeat-clinic-ios
//
//  Created by Manh Tran on 17/06/20.
//  Copyright © 2020 itr.vn. All rights reserved.
//

import Foundation

enum CHANNEL: String {
    case CH_1
    case CH_2
    case CH_3
}

class NotifyECGDataHandler: NotifyHandler {
    public static func handle(packet: Packet, blueHandler: BluetoothHandler, callback: DXHDeviceHandlerCallback?) {
        
        guard !isResponseFailed(bluetoothHandler: blueHandler, packet: packet) else {
            return
        }
        
        if  let values = unPackPacket(packet: packet, validCount: 3) {
            //        let index = unpackValue?.values[1].uintValue
            if let data = (values[2].dataValue) {
                callback?.updateECGData(data: data, bluetoothHandler: blueHandler)
            }
        }
    }
    
    public static func parseECGSample(ECGConfig: ECGConfig, data: Data) -> ECGSample? {
        let channels = ECGConfig.convertChannel()
        let gain = ECGConfig.gain
        switch channels.count {
        case 1:
            return parseOneChannel(channels: channels, gain: gain, data: data)
        case 2:
            return parseTwoChannels(channels: channels, gain: gain, data: data)
        case 3:
            return parseThreeChannels(channels: channels, gain: gain, data: data)
        default:
            return nil
        }
    }
    
    private static func isResponseFailed(bluetoothHandler: BluetoothHandler?, packet: Packet) -> Bool {
        guard bluetoothHandler!.isConnected else {
            return true
        }
        return false
    }
    
    private static func parseOneChannel(channels: [CHANNEL], gain: Double, data: Data) -> ECGSample? {
        var ch = [Int16]()
        for i in stride(from: 0, to: data.count, by: 2) {
            let sample = data.sub(in: i...i+2)
            guard let value = sample?.toInt16(byteOrder: .LittleEndian) else { return nil}
            ch.append(value)
        }
        
        switch channels[0] {
        case .CH_1:
            return ECGSample(ch1: ch, ch2: nil, ch3: nil)
        case .CH_2:
            return ECGSample(ch1: ch, ch2: nil, ch3: nil)
        case .CH_3:
            return ECGSample(ch1: ch, ch2: nil, ch3: nil)
        }
    }
    
    private static func parseTwoChannels(channels: [CHANNEL], gain: Double, data: Data) -> ECGSample? {
        var data1 = [Int16]()
        var data2 = [Int16]()
        for i in stride(from: 0, to: data.count, by: 4) {
            let sample1 = data.sub(in: i...i+2)
            let sample2 = data.sub(in: i+2...i+4)
            guard let value1 = sample1?.toInt16(byteOrder: .LittleEndian) else { return nil}
            guard let value2 = sample2?.toInt16(byteOrder: .LittleEndian) else { return nil}
            data1.append(value1)
            data2.append(value2)
        }
        let chValue1 = channels[0]
        let chValue2 = channels[1]
        if(chValue1 == .CH_1 && chValue2 == .CH_2) {
            return ECGSample(ch1: data1, ch2: data2, ch3: nil)
        } else if (chValue1 == .CH_1 && chValue2 == .CH_3) {
            return ECGSample(ch1: data1, ch2: nil, ch3: data2)
        } else {
            return ECGSample(ch1: nil, ch2: data1, ch3: data2)
        }
    }
    
    private static func parseThreeChannels(channels: [CHANNEL], gain: Double, data: Data) -> ECGSample? {
        var data1 = [Int16]()
        var data2 = [Int16]()
        var data3 = [Int16]()
        for i in stride(from: 0, to: data.count, by: 6) {
            let sample1 = data.sub(in: i...i+2)
            let sample2 = data.sub(in: i+2...i+4)
            let sample3 = data.sub(in: i+4...i+6)
            guard let value1 = sample1?.toInt16(byteOrder: .LittleEndian) else { return nil}
            guard let value2 = sample2?.toInt16(byteOrder: .LittleEndian) else { return nil}
            guard let value3 = sample3?.toInt16(byteOrder: .LittleEndian) else { return nil}
            data1.append(value1)
            data2.append(value2)
            data3.append(value3)
        }
        return ECGSample(ch1: data1, ch2: data2, ch3: data3)
    }
    
}

File: plugin/octo_beat_plugin/android/src/main/kotlin/com/octo/octo_beat_plugin/core/device/handler/NotifyECGDataHandler.kt
package com.octo.octo_beat_plugin.core.device.handler

import android.util.Log
import com.octo.octo_beat_plugin.core.device.DeviceHandlerCallback
import com.octo.octo_beat_plugin.core.device._enum.CHANNEL
import com.octo.octo_beat_plugin.core.device.model.DXHDevice
import com.octo.octo_beat_plugin.core.device.model.ECGConfig
import com.octo.octo_beat_plugin.core.device.model.ECGSample
import com.octo.octo_beat_plugin.core.utils.ByteUtils
import com.octo.octo_beat_plugin.core.utils.DeviceUtils
import com.octo.octo_beat_plugin.core.utils.MessagePackHelper
import java.nio.ByteOrder
import java.io.File
import java.io.FileOutputStream
import java.io.OutputStreamWriter


class NotifyECGDataHandler {
    companion object {
        private val TAG = "NotifyECGDataHandler"
        private var previousIndex = 0u
        private var lost = 0u
//        val bufferString: StringBuffer = StringBuffer(100*1024)


        fun handle(dxhDevice: DXHDevice, payloadData: ByteArray, deviceHandlerCallback: DeviceHandlerCallback) {
            val value = MessagePackHelper.unpackRaw(payloadData)
            val index = value.asArrayValue().get(1).asIntegerValue().toInt().toUInt()
//            Log.d(TAG, "indexECG: $index")
//            checkLostPacketIndex(index)
            val bin = value.asArrayValue().get(2).asBinaryValue().asByteArray()
            saveBinTobuffer(ByteUtils.toHexString(bin))
            dxhDevice.ECGData = bin
            deviceHandlerCallback.updateECGData(dxhDevice)
        }

        fun parseSampleECG(ECGConfig: ECGConfig, data: ByteArray): ECGSample? {
            val channels = DeviceUtils.convertChannel(ECGConfig.channel)
            val gain = ECGConfig.gain
            return when (channels.size) {
                1 -> parseOneChannel(channels, gain, data)
                2 -> parseTwoChannel(channels, gain, data)
                3 -> parseThreeChannel(channels, gain, data)
                else -> null
            }
        }

        private fun parseOneChannel(channels: ArrayList<CHANNEL>, gain: Double, data: ByteArray): ECGSample {
            val ch = ArrayList<Short>()
            for (i in 0..data.size - 2 step 2) {
                val sample = ByteUtils.subByteArray(data, i, 2)
                val value = ByteUtils.toShort(sample, ByteOrder.LITTLE_ENDIAN)
                ch.add(value)
            }
            return when (channels[0]) {
                CHANNEL.CH_1 -> ECGSample(ch, null, null)
                CHANNEL.CH_2 -> ECGSample(null, ch, null)
                CHANNEL.CH_3 -> ECGSample(null, null, ch)
            }
        }

        private fun parseTwoChannel(channels: ArrayList<CHANNEL>, gain: Double, data: ByteArray): ECGSample? {
            val data1 = ArrayList<Short>()
            val data2 = ArrayList<Short>()
            for (i in 0..data.size - 4 step 4) {
                val sample1 = ByteUtils.subByteArray(data, i, 2)
                val sample2 = ByteUtils.subByteArray(data, i + 2, 2)
                val value1 = ByteUtils.toShort(sample1, ByteOrder.LITTLE_ENDIAN)
                val value2 = ByteUtils.toShort(sample2, ByteOrder.LITTLE_ENDIAN)
                data1.add(value1)
                data2.add(value2)
            }

            val chValue1 = channels[0]
            val chValue2 = channels[1]
            return if (chValue1 == CHANNEL.CH_1 && chValue2 == CHANNEL.CH_2) {
                ECGSample(data1, data2, null)
            } else if (chValue1 == CHANNEL.CH_1 && chValue2 == CHANNEL.CH_3) {
                ECGSample(data1, null, data2)
            } else {
                ECGSample(null, data1, data2)
            }
        }

        private  fun  parseThreeChannel(channels: ArrayList<CHANNEL>, gain: Double, data: ByteArray): ECGSample? {
            val data1 = ArrayList<Short>()
            val data2 = ArrayList<Short>()
            val data3 = ArrayList<Short>()

//            Log.d(TAG, "data size: ${data.size}: ${ByteUtils.toHexString(data)}")
            for (i in 0..data.size - 6 step 6) {
                val sample1 = ByteUtils.subByteArray(data, i, 2)
                val sample2 = ByteUtils.subByteArray(data, i + 2, 2)
                val sample3 = ByteUtils.subByteArray(data, i + 4, 2)
                val value1 = ByteUtils.toShort(sample1, ByteOrder.LITTLE_ENDIAN)
                val value2 = ByteUtils.toShort(sample2, ByteOrder.LITTLE_ENDIAN)
                val value3 = ByteUtils.toShort(sample3, ByteOrder.LITTLE_ENDIAN)
                data1.add(value1)
                data2.add(value2)
                data3.add(value3)
//                Log.d(TAG, "data CH1: $value1")
            }

//            Log.d(TAG, "ch1 size: ${data1.size} ch2 size: ${data2.size} ch3 size: ${data3.size}")
            return ECGSample(data1, data2, data3)
        }

        fun checkLostPacketIndex(currentIndex: UInt) {
            lost = currentIndex - previousIndex
//            Log.d(TAG, "lost indexECG: $lost")
            if (lost < 0u) {
                previousIndex = 0u
                lost = currentIndex - previousIndex
            }
            if (lost > 0u && lost != 1u) {
                Log.d(TAG, "lost: $lost")
            }
            previousIndex = currentIndex
        }

        fun saveBinTobuffer(data: String) {
            var value: String = data.replace("]", "").replace("[", " ,")
//            bufferString.append(value)
        }

        fun writeSDcard() {
            try {
                val myFile = File("/sdcard/mysdfile.txt")
                myFile.createNewFile()
                val fOut = FileOutputStream(myFile)
                val myOutWriter = OutputStreamWriter(fOut)
//                myOutWriter.append(bufferString)
                myOutWriter.close()
                fOut.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

    }
}